src/query_parser/v2/parser.y

```diff
@@ -445,78 +443,35 @@ text_union(A) ::= text_union(B) OR text_expr(C). [OR] {
 /////////////////////////////////////////////////////////////////
 
 expr(A) ::= modifier(B) COLON text_expr(C) . {
-    if (C == NULL) {
-        A = NULL;
-    } else {
-        if (ctx->sctx->spec) {
-            QueryNode_SetFieldMask(C, IndexSpec_GetFieldBit(ctx->sctx->spec, B.s, B.len));
-        }
-        A = C;
+  if (ctx->sctx->spec && !FIELD_IS(B.fs, INDEXFLD_T_FULLTEXT)) {
+    REPORT_WRONG_FIELD_TYPE(B, SPEC_TEXT_STR);
+    QueryNode_Free(C);
+    A = NULL;
+  } else if (C == NULL) {
+    A = NULL;
+  } else {
+    if (ctx->sctx->spec) {
+      QueryNode_SetFieldMask(C, FIELD_BIT(B.fs));
     }
-}
-
-expr(A) ::= modifier(B) NOT_EQUAL param_num(C) . {
-  QueryParam *qp = NewNumericFilterQueryParam_WithParams(ctx, &C, &C, 1, 1);
-  qp->nf->fieldName = rm_strndup(B.s, B.len);
-  QueryNode* E = NewNumericNode(qp);
-  A = NewNotNode(E);
-}
-
-expr(A) ::= modifier(B) EQUALS param_num(C) . {
-  QueryParam *qp = NewNumericFilterQueryParam_WithParams(ctx, &C, &C, 1, 1);
-  qp->nf->fieldName = rm_strndup(B.s, B.len);
-  A = NewNumericNode(qp);
-}
-
-expr(A) ::= modifier(B) GT param_num(C) . {
-  QueryParam *qp = NewNumericFilterQueryParam_WithParams(ctx, &C, NULL, 0, 1);
-  qp->nf->fieldName = rm_strndup(B.s, B.len);
-  A = NewNumericNode(qp);
-}
-
-expr(A) ::= modifier(B) GE param_num(C) . {
-  QueryParam *qp = NewNumericFilterQueryParam_WithParams(ctx, &C, NULL, 1, 1);
-  qp->nf->fieldName = rm_strndup(B.s, B.len);
-  A = NewNumericNode(qp);
-}
-
-expr(A) ::= modifier(B) LT param_num(C) . {
-  QueryParam *qp = NewNumericFilterQueryParam_WithParams(ctx, NULL, &C, 1, 0);
-  qp->nf->fieldName = rm_strndup(B.s, B.len);
-  A = NewNumericNode(qp);
-}
-
-expr(A) ::= modifier(B) LE param_num(C) . {
-  QueryParam *qp = NewNumericFilterQueryParam_WithParams(ctx, NULL, &C, 1, 1);
-  qp->nf->fieldName = rm_strndup(B.s, B.len);
-  A = NewNumericNode(qp);
+    A = C;
+  }
 }
 
 expr(A) ::= modifierlist(B) COLON text_expr(C) . {
-
-    if (C == NULL) {
-        for (size_t i = 0; i < Vector_Size(B); i++) {
-          char *s;
-          Vector_Get(B, i, &s);
-          rm_free(s);
-        }
-        Vector_Free(B);
-        A = NULL;
-    } else {
-        //C->opts.fieldMask = 0;
-        t_fieldMask mask = 0;
-        for (int i = 0; i < Vector_Size(B); i++) {
-            char *p;
-            Vector_Get(B, i, &p);
-            if (ctx->sctx->spec) {
-              mask |= IndexSpec_GetFieldBit(ctx->sctx->spec, p, strlen(p));
-            }
-            rm_free(p);
-        }
-        Vector_Free(B);
-        QueryNode_SetFieldMask(C, mask);
-        A=C;
+  if (C == NULL) {
+    array_free(B);
+    A = NULL;
+  } else {
+    t_fieldMask mask = 0;
+    if (ctx->sctx->spec) {
+      for (int i = 0; i < array_len(B); i++) {
+        mask |= FIELD_BIT(B[i].fs);
+      }
     }
+    array_free(B);
```

Comment:
Is this needed given the defined destructor?

Reply:
Yes. See the [destructor](https://sqlite.org/src/doc/trunk/doc/lemon.html#destructor) section. Mainly:

> It is important to note that the value of a non-terminal is passed to the destructor whenever the non-terminal is removed from the stack, unless the non-terminal is used in a C-code action. If the non-terminal is used by C-code, then it is assumed that the C-code will take care of destroying it. More commonly, the value is used to build some larger structure, and we don't want to destroy it, which is why the destructor is not called in this circumstance.